viv_frontend.py

```diff
@@ -0,0 +1,672 @@
+            state.structs_lock.release()
+            '''
+            vw.vprint('revsync: <%s> %s %s %s' % (user, cmd, old_struct_name, new_struct_name))
+        
+        elif cmd == 'struc_member_created':
+            struct_name = data['struc_name'].encode('ascii', 'ignore')
+            '''  not ready to wrap this into Viv yet.
+            state.structs_lock.acquire()
+            struct = bv.get_type_by_name(struct_name)
+            if struct is None:
+                vw.vprint('revsync: unknown struct name %s during struc_member_created cmd' % struct_name)
+                return
+            member_name = data['member_name'].encode('ascii', 'ignore')
+            struct_type = get_type_by_size(bv, data['size'])
+            if struct_type is None:
+                vw.vprint('revsync: bad struct member size %d for member %s during struc_member_created cmd' % (data['size'], member_name))
+                return
+            # need actual Structure class, not Type
+            struct = struct.structure
+            struct.insert(data['offset'], struct_type, member_name)
+            # we must redefine the type
+            bv.define_user_type(struct_name, binaryninja.types.Type.structure_type(struct))
+            state.structs = get_structs(bv)
+            state.structs_lock.release()
+            '''
+            vw.vprint('revsync: <%s> %s %s->%s' % (user, cmd, struct_name, member_name))
+        
+        elif cmd == 'struc_member_deleted':
+            struct_name = data['struc_name'].encode('ascii', 'ignore')
+            member_name = '???'
+            '''  not ready to wrap this into Viv yet.
+            state.structs_lock.acquire()
+            struct = bv.get_type_by_name(struct_name)
+            if struct is None:
+                vw.vprint('revsync: unknown struct name %s during struc_member_deleted cmd' % struct_name)
+                return
+            offset = data['offset']
+            # need actual Structure class, not Type
+            struct = struct.structure
+            # walk the list and find the index to delete (seriously, why by index binja and not offset?)
+            member_name = '???'
+            for i,m in enumerate(struct.members):
+                if m.offset == offset:
+                    # found it
+                    member_name = m.name
+                    struct.remove(i)
+            # we must redefine the type
+            bv.define_user_type(struct_name, binaryninja.types.Type.structure_type(struct))
+            state.structs = get_structs(bv)
+            state.structs_lock.release()
+            '''
+            vw.vprint('revsync: <%s> %s %s->%s' % (user, cmd, struct_name, member_name))
+            
+        elif cmd == 'struc_member_renamed':
+            struct_name = data['struc_name'].encode('ascii', 'ignore')
+            member_name = data['member_name'].encode('ascii', 'ignore')
+            '''  not ready to wrap this into Viv yet.
```

Comment:
can you remove these multiline strings? there's no reason to duplicate the binja code in a string

Reply:
done.